### ISSUE TITLE
Generating pages from json file increases the size of bundle

### ISSUE THREAD

**Issue by tareefdev** [gatsbyjs/gatsby#14455]
## Description
Using a JSON file to generate HTML static pages results in a big JS bundle size.

### Steps to reproduce
- Clone the gatsby-starter-blog repo.
- Add JSON file to it (I used a [file](https://github.com/tareefdev/test-json/tree/master/src/data) with ~16MB of size)
- [Query](https://github.com/tareefdev/test-json/blob/master/gatsby-node.js#L28) it using GraphQL in `gatsby-node`
- Add a template [file](https://github.com/tareefdev/test-json/blob/master/src/templates/unit-post.js) for the wanted pages.
- Build the site.

### Expected result
The size of JS bundle should not grow as I didn't add any new JavaScript library.

### Actual result
The bundle size jumped from ~300KB to ~3.2MB for new 5K pages, and to ~7.2MB for 12K page.

### Environment
````
System:
    OS: Linux 5.1 Arch Linux undefined
    CPU: (4) x64 Intel(R) Core(TM) i7-6500U CPU @ 2.50GHz
    Shell: 5.0.7 - /bin/bash
  Binaries:
    Node: 8.16.0 - ~/.nvm/versions/node/v8.16.0/bin/node
    npm: 6.4.1 - ~/.nvm/versions/node/v8.16.0/bin/npm
  Languages:
    Python: 3.7.3 - /usr/bin/python
  Browsers:
    Firefox: 67.0
  npmPackages:
    gatsby: ^2.7.5 => 2.7.5
    gatsby-image: ^2.1.2 => 2.1.2
    gatsby-plugin-feed: ^2.2.1 => 2.2.1
    gatsby-plugin-google-analytics: ^2.0.20 => 2.0.20
    gatsby-plugin-manifest: ^2.1.1 => 2.1.1
    gatsby-plugin-offline: ^2.1.1 => 2.1.1
    gatsby-plugin-react-helmet: ^3.0.12 => 3.0.12
    gatsby-plugin-sharp: ^2.1.2 => 2.1.2
    gatsby-plugin-typography: ^2.2.13 => 2.2.13
    gatsby-remark-copy-linked-files: ^2.0.13 => 2.0.13
    gatsby-remark-images: ^3.0.14 => 3.0.14
    gatsby-remark-prismjs: ^3.2.9 => 3.2.9
    gatsby-remark-responsive-iframe: ^2.1.1 => 2.1.1
    gatsby-remark-smartypants: ^2.0.9 => 2.0.9
    gatsby-source-filesystem: ^2.0.37 => 2.0.37
    gatsby-transformer-json: ^2.1.11 => 2.1.11
    gatsby-transformer-remark: ^2.3.12 => 2.3.12
    gatsby-transformer-sharp: ^2.1.20 => 2.1.20
  npmGlobalPackages:
    gatsby-cli: 2.6.3
    gatsby: 2.4.2
````

**Comment by pieh** [comment]
Right now we do bundle pages manifest (list of all pages, what template they use and link to results of the query) which is main growing chunk. There is ongoing work of changing that part - https://github.com/gatsbyjs/gatsby/pull/14359 that will make Gatsby sites more scalable. You can try it using `npm install gatsby@per-page-manifest`

**Comment by tareefdev** [comment]
That's great! The bundle size decreased from 7.2MB to 400KB on the homepage. Should I stay on this branch or it merge shortly into master?

**Comment by pieh** [comment]
Hope is to merge it soon-ish, but given size of the change it might take a while to merge. I would keep using dist-tag for now and maybe subscribe to linked pull request to keep track when it's merged

**Comment by tareefdev** [comment]
Thanks for helping, also for @Moocar for his amazing work!

**Comment by KyleAMathews** [comment]
@moocar's PR is now published in 2.9.0 🎉

**Comment by tareefdev** [comment]
great!! Thanks
